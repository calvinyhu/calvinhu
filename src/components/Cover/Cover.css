@-webkit-keyframes fadeIn{0%{opacity:0}100%{opacity:1}}@-moz-keyframes fadeIn{0%{opacity:0}100%{opacity:1}}@-o-keyframes fadeIn{0%{opacity:0}100%{opacity:1}}@keyframes fadeIn{0%{opacity:0}100%{opacity:1}}@-webkit-keyframes fadeIn0to75{0%{opacity:0}100%{opacity:.75}}@-moz-keyframes fadeIn0to75{0%{opacity:0}100%{opacity:.75}}@-o-keyframes fadeIn0to75{0%{opacity:0}100%{opacity:.75}}@keyframes fadeIn0to75{0%{opacity:0}100%{opacity:.75}}@-webkit-keyframes fadeOut75to0{0%{opacity:.75}100%{opacity:0}}@-moz-keyframes fadeOut75to0{0%{opacity:.75}100%{opacity:0}}@-o-keyframes fadeOut75to0{0%{opacity:.75}100%{opacity:0}}@keyframes fadeOut75to0{0%{opacity:.75}100%{opacity:0}}@-webkit-keyframes blockSlideY{0%{-webkit-transform:translateY(5rem);transform:translateY(5rem)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}@-moz-keyframes blockSlideY{0%{-moz-transform:translateY(5rem);transform:translateY(5rem)}100%{-moz-transform:translateY(0);transform:translateY(0)}}@-o-keyframes blockSlideY{0%{-o-transform:translateY(5rem);transform:translateY(5rem)}100%{-o-transform:translateY(0);transform:translateY(0)}}@keyframes blockSlideY{0%{-webkit-transform:translateY(5rem);-moz-transform:translateY(5rem);-o-transform:translateY(5rem);transform:translateY(5rem)}100%{-webkit-transform:translateY(0);-moz-transform:translateY(0);-o-transform:translateY(0);transform:translateY(0)}}@-webkit-keyframes slideYOut5{0%{-webkit-transform:translateY(0);transform:translateY(0)}100%{-webkit-transform:translateY(5rem);transform:translateY(5rem)}}@-moz-keyframes slideYOut5{0%{-moz-transform:translateY(0);transform:translateY(0)}100%{-moz-transform:translateY(5rem);transform:translateY(5rem)}}@-o-keyframes slideYOut5{0%{-o-transform:translateY(0);transform:translateY(0)}100%{-o-transform:translateY(5rem);transform:translateY(5rem)}}@keyframes slideYOut5{0%{-webkit-transform:translateY(0);-moz-transform:translateY(0);-o-transform:translateY(0);transform:translateY(0)}100%{-webkit-transform:translateY(5rem);-moz-transform:translateY(5rem);-o-transform:translateY(5rem);transform:translateY(5rem)}}.Hide{opacity:0}.Show{opacity:1}.FadeIn{-webkit-animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards;-moz-animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards;-o-animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards;animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards}.FabSlideFadeIn{-webkit-animation:fadeIn0to75 .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards;-moz-animation:fadeIn0to75 .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards;-o-animation:fadeIn0to75 .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards;animation:fadeIn0to75 .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards}.FabSlideFadeOut{-webkit-animation:fadeOut75to0 .5s cubic-bezier(.26,.94,.58,1) forwards,slideYOut5 .5s cubic-bezier(.26,.94,.58,1) forwards;-moz-animation:fadeOut75to0 .5s cubic-bezier(.26,.94,.58,1) forwards,slideYOut5 .5s cubic-bezier(.26,.94,.58,1) forwards;-o-animation:fadeOut75to0 .5s cubic-bezier(.26,.94,.58,1) forwards,slideYOut5 .5s cubic-bezier(.26,.94,.58,1) forwards;animation:fadeOut75to0 .5s cubic-bezier(.26,.94,.58,1) forwards,slideYOut5 .5s cubic-bezier(.26,.94,.58,1) forwards}.BlockSlideFadeIn{-webkit-animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards;-moz-animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards;-o-animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards;animation:fadeIn .5s cubic-bezier(.26,.94,.58,1) forwards,blockSlideY .5s cubic-bezier(.26,.94,.58,1) forwards}@-webkit-keyframes fadeOut{0%{opacity:1}100%{opacity:0}}@-moz-keyframes fadeOut{0%{opacity:1}100%{opacity:0}}@-o-keyframes fadeOut{0%{opacity:1}100%{opacity:0}}@keyframes fadeOut{0%{opacity:1}100%{opacity:0}}@-webkit-keyframes textSlideY{0%{-webkit-transform:translateY(50%);transform:translateY(50%)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}@-moz-keyframes textSlideY{0%{-moz-transform:translateY(50%);transform:translateY(50%)}100%{-moz-transform:translateY(0);transform:translateY(0)}}@-o-keyframes textSlideY{0%{-o-transform:translateY(50%);transform:translateY(50%)}100%{-o-transform:translateY(0);transform:translateY(0)}}@keyframes textSlideY{0%{-webkit-transform:translateY(50%);-moz-transform:translateY(50%);-o-transform:translateY(50%);transform:translateY(50%)}100%{-webkit-transform:translateY(0);-moz-transform:translateY(0);-o-transform:translateY(0);transform:translateY(0)}}.Cover{display:-webkit-box;display:-webkit-flex;display:-moz-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-webkit-flex-direction:column;-moz-box-orient:vertical;-moz-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:center;-webkit-justify-content:center;-moz-box-pack:center;-ms-flex-pack:center;justify-content:center;-webkit-box-align:center;-webkit-align-items:center;-moz-box-align:center;-ms-flex-align:center;align-items:center;background:#323232;color:#000;position:relative;width:100%;height:100vh;overflow:hidden}.Cover .Background{position:absolute;width:100%;height:100vh;background:url(../../assets/images/DSC_9569.jpg);-webkit-background-size:cover;-moz-background-size:cover;-o-background-size:cover;background-size:cover;background-attachment:fixed;background-position:center left 25%;opacity:0;-webkit-animation:fadeIn 1s linear .5s forwards;-moz-animation:fadeIn 1s linear .5s forwards;-o-animation:fadeIn 1s linear .5s forwards;animation:fadeIn 1s linear .5s forwards}.Cover .Blur{position:absolute;width:250vw;height:100vh;-webkit-border-radius:50%;-moz-border-radius:50%;border-radius:50%;background:-webkit-radial-gradient(rgba(255,255,255,.8) 0,transparent 35%);background:-moz-radial-gradient(rgba(255,255,255,.8) 0,transparent 35%);background:-o-radial-gradient(rgba(255,255,255,.8) 0,transparent 35%);background:radial-gradient(rgba(255,255,255,.8) 0,transparent 35%);opacity:0;-webkit-animation:fadeIn 1s linear .5s forwards;-moz-animation:fadeIn 1s linear .5s forwards;-o-animation:fadeIn 1s linear .5s forwards;animation:fadeIn 1s linear .5s forwards}.SocialMedia{display:-webkit-box;display:-webkit-flex;display:-moz-box;display:-ms-flexbox;display:flex;-webkit-box-pack:end;-webkit-justify-content:flex-end;-moz-box-pack:end;-ms-flex-pack:end;justify-content:flex-end;position:absolute;top:0;width:100%;padding:1rem}.SocialMedia h5{margin:0;margin-left:1rem}@media (min-width:40rem){.SocialMedia{padding:2rem}}.CoverText{display:-webkit-box;display:-webkit-flex;display:-moz-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-webkit-flex-direction:column;-moz-box-orient:vertical;-moz-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:center;-webkit-justify-content:center;-moz-box-pack:center;-ms-flex-pack:center;justify-content:center;-webkit-box-align:center;-webkit-align-items:center;-moz-box-align:center;-ms-flex-align:center;align-items:center;width:85%}@media (min-width:40rem){.CoverText{-webkit-box-orient:horizontal;-webkit-box-direction:normal;-webkit-flex-direction:row;-moz-box-orient:horizontal;-moz-box-direction:normal;-ms-flex-direction:row;flex-direction:row;-webkit-justify-content:space-around;-ms-flex-pack:distribute;justify-content:space-around}}.Blurb{display:-webkit-box;display:-webkit-flex;display:-moz-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-webkit-flex-direction:column;-moz-box-orient:vertical;-moz-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:center;-webkit-justify-content:center;-moz-box-pack:center;-ms-flex-pack:center;justify-content:center;-webkit-box-align:center;-webkit-align-items:center;-moz-box-align:center;-ms-flex-align:center;align-items:center}.Blurb .Name,.Blurb .Title{margin-bottom:1rem;text-align:center;-webkit-transform:translateY(100%);-moz-transform:translateY(100%);-ms-transform:translateY(100%);-o-transform:translateY(100%);transform:translateY(100%);opacity:0;-webkit-animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) .5s forwards,textSlideY 1s cubic-bezier(.26,.94,.58,1) .5s forwards;-moz-animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) .5s forwards,textSlideY 1s cubic-bezier(.26,.94,.58,1) .5s forwards;-o-animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) .5s forwards,textSlideY 1s cubic-bezier(.26,.94,.58,1) .5s forwards;animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) .5s forwards,textSlideY 1s cubic-bezier(.26,.94,.58,1) .5s forwards}@media (min-width:40rem){.Blurb .Name,.Blurb .Title{font-size:4rem;margin-bottom:0}}@media (min-width:40rem){.Blurb .Title{font-size:3rem;margin-bottom:0}}.Nav{display:-webkit-box;display:-webkit-flex;display:-moz-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-webkit-flex-direction:column;-moz-box-orient:vertical;-moz-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:center;-webkit-justify-content:center;-moz-box-pack:center;-ms-flex-pack:center;justify-content:center;-webkit-box-align:center;-webkit-align-items:center;-moz-box-align:center;-ms-flex-align:center;align-items:center}.Nav h5{margin:.5rem 0;opacity:0;-webkit-animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) 2s forwards;-moz-animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) 2s forwards;-o-animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) 2s forwards;animation:fadeIn 1s cubic-bezier(.26,.94,.58,1) 2s forwards}.Nav h5:hover{cursor:pointer;color:#fff}@media (min-width:40rem){.Nav{-webkit-box-align:start;-webkit-align-items:flex-start;-moz-box-align:start;-ms-flex-align:start;align-items:flex-start}}