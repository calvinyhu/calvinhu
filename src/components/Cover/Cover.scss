@import './src/assets/scss/_variables.scss';
@import './src/assets/scss/_mixins.scss';
@import './src/assets/scss/_classes.scss';

@include fade-out;
@include slide-y(textSlideY, 50%, 0%);

.Cover {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: $dark;
  color: black;
  position: relative;
  width: 100%;
  height: 100vh;
  overflow: hidden;
  white-space: nowrap;

  .Background {
    position: absolute;
    width: 100%;
    height: 100vh;
    background: url('../../assets/images/DSC_9569.jpg');
    background-size: cover;
    background-attachment: fixed;
    background-position: center left 25%;
    opacity: 0;
    animation: fadeIn 1s $timing-func-2 0.5s forwards;
  }

  .Blur {
    position: absolute;
    width: 250vw;
    height: 100vh;
    border-radius: 50%;
    background: radial-gradient(rgba(255, 255, 255, 0.8) 0%, transparent 35%);
    opacity: 0;
    animation: fadeIn 1s $timing-func-2 0.5s forwards;
  }
}

.CoverText {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 85%;

  @include media-min-width($desktop-width) {
    flex-direction: row;
    justify-content: space-around;
  }
}

.Blurb {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  .Name,
  .Title {
    margin-bottom: $margin;
    text-align: center;
    transform: translateY(100%);
    opacity: 0;
    animation: fadeIn 1.5s $timing-func-1 1s forwards,
      textSlideY 1s $timing-func-1 1s forwards;

    @include media-min-width($desktop-width) {
      font-size: $font-size * 4;
    }
  }

  .Title {
    @include media-min-width($desktop-width) {
      font-size: $font-size * 3;
      margin-bottom: 0;
    }
  }
}

.Nav {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: rgba(255, 255, 255, 0.3);
  padding: $padding $padding * 2;
  border-radius: $border-radius-card;
  margin-top: $margin * 2;
  opacity: 0;
  animation: fadeIn 1s $timing-func-1 2s forwards;

  @include media-min-width($desktop-width) {
    align-items: flex-start;
    margin-top: 0;
    margin-left: $margin * 2;
  }

  h5 {
    margin: $margin / 2 0;
    transition: $transition2;

    &:hover {
      cursor: pointer;
      color: $dark-opp-color;
      transform: scale(1.2);
    }
  }

  .SocialMedia {
    display: flex;
    justify-content: space-between;
    width: 100%;
    margin: $margin / 2 0;

    .ImgContainer {
      width: 2rem;
      transition: $transition2;

      &:hover {
        cursor: pointer;
        transform: scale(1.2);
      }

      img {
        width: 100%;
        height: 100%;
        object-fit: contain;
        object-position: center left;
      }
    }
  }
}
