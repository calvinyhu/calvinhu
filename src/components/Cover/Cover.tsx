import React, { useCallback, useEffect } from 'react';
import { useSpring, animated, interpolate } from 'react-spring';

import Fa from '../UI/Fa/Fa';
import Button from '../UI/Button/Button';
import packageJson from '../../../package.json';

import styles from './Cover.module.scss';
import p1080 from '../../assets/images/DSC_9569-1080p50-blurred.jpg';
import p1440 from '../../assets/images/DSC_9569-1440p50-blurred.jpg';
import p2160 from '../../assets/images/DSC_9569-2160p35-blurred.jpg';

const Cover = () => {
  const [{ st, xy }, setSpring] = useSpring(() => ({
    st: 0,
    xy: [0, 0],
    config: { mass: 1, tension: 1000, friction: 50 },
  }));

  const handleMouseMove = useCallback(
    ({ clientX, clientY }) =>
      setSpring({
        xy: [clientX - window.innerWidth / 2, clientY - window.innerHeight / 2],
      }),
    [setSpring],
  );

  const onScroll = useCallback(() => setSpring({ st: window.scrollY / 4 }), [
    setSpring,
  ]);

  useEffect(() => {
    const event = 'scroll';
    window.addEventListener(event, onScroll);

    return () => {
      window.removeEventListener(event, onScroll);
    };
  });

  // @ts-ignore
  const interpolateBackground = interpolate([st, xy], (o, xy) => {
    // @ts-ignore
    return `translate3d(${-xy[0] / 15}px, ${-xy[1] / 15 -
      o}px, 0px) scale3d(1.4,1.4,1.4)`;
  });

  const handleScrollDown = () =>
    window.scrollTo({ top: window.innerHeight + 1 });

  return (
    <div className={styles.Cover} onMouseMove={handleMouseMove}>
      <animated.div
        style={{ transform: (interpolateBackground as unknown) as string }}
        className={styles.Background}
      >
        <picture>
          <source media="(min-width: 1440px)" srcSet={p1440} />
          <source media="(min-width: 2160px)" srcSet={p2160} />
          <img src={p1080} alt="me" />
        </picture>
        <div className={styles.LightenFilter} />
      </animated.div>

      <div className={styles.CoverContent}>
        <div className={styles.Text}>
          <h1>Calvin</h1>
          <h1>Hu</h1>
        </div>
        <div className={styles.Icon}>
          <Button circle clear click={handleScrollDown} ariaLabel="Scroll Down">
            <Fa lg>fas fa-angle-down</Fa>
          </Button>
        </div>
        <div className={styles.version}>{packageJson.version}</div>
      </div>
    </div>
  );
};

export default Cover;
