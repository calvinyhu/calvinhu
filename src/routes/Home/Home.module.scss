@import './src/styles/_variables.scss';
@import './src/styles/_mixins.scss';
@import './src/styles/_classes.scss';

.Home {
  width: 100%;
  height: 100%;
}
