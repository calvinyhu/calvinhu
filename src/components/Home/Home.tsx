import React, { useEffect, useState, FC } from 'react';
// @ts-ignore
import throttle from 'raf-throttle';

import Cover from 'components/Cover/Cover';
import Milestone from 'components/Milestone/Milestone';
import { useResetScrollOnUnmount } from 'utils/hooks';

import milestones from 'constants/milestones';

interface HomeProps {
  path: string;
}

const Home: FC<HomeProps> = () => {
  const [offsetX, setOffsetX] = useState(0);

  useEffect(() => {
    const animatePage = (scrollTop: number, clientHeight: number) => {
      if (scrollTop < clientHeight) setOffsetX(-scrollTop / 10);
    };

    const handleScroll = () => {
      throttle(animatePage(window.pageYOffset, window.innerHeight));
    };

    const event = 'scroll';
    window.addEventListener(event, handleScroll);

    return () => {
      window.removeEventListener(event, handleScroll);
    };
  });

  useResetScrollOnUnmount();

  return (
    <>
      <Cover offsetX={offsetX} />
      {Object.values(milestones).map(
        ({ title, alt, date, description, photo, url, urlTitle }) => (
          <Milestone
            key={title}
            alt={alt}
            date={date}
            description={description}
            photo={photo}
            title={title}
            url={url}
            urlTitle={urlTitle}
          />
        ),
      )}
    </>
  );
};

export default Home;
