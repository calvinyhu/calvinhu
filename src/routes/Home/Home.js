import React, { useEffect, useState } from 'react';
import throttle from 'raf-throttle';

import styles from './Home.module.scss';
import milestones from '../../assets/milestones/milestones';
import Cover from '../../components/Cover/Cover';
import Milestone from '../../components/Milestone/Milestone';

const Home = () => {
  const [offsetX, setOffsetX] = useState(0);

  useEffect(() => {
    const animatePage = (scrollTop, clientHeight) => {
      if (scrollTop < clientHeight) setOffsetX(-scrollTop / 10);
    };

    const handleScroll = () => {
      throttle(animatePage(window.pageYOffset, window.innerHeight));
    };

    const event = 'scroll';
    window.addEventListener(event, handleScroll);

    return () => {
      window.removeEventListener(event, handleScroll);
    };
  });

  const renderMilestones = () =>
    Object.values(milestones).map(val => (
      <Milestone
        key={val.title}
        alt={val.alt}
        date={val.date}
        description={val.description}
        photo={val.photo}
        title={val.title}
        url={val.url}
        urlTitle={val.urlTitle}
      />
    ));

  return (
    <div className={styles.Home}>
      <Cover offsetX={offsetX} />
      {renderMilestones()}
    </div>
  );
};

export default Home;
