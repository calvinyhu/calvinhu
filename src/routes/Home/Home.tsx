import React, { useEffect, useState } from 'react';
// @ts-ignore
import throttle from 'raf-throttle';

import Cover from 'components/Cover/Cover';
import Milestone from 'components/Milestone/Milestone';
import { useResetScrollOnUnmount } from 'utils/hooks';

import styles from './Home.module.scss';
import milestones from 'assets/milestones/milestones';

const Home = () => {
  const [offsetX, setOffsetX] = useState(0);

  useEffect(() => {
    const animatePage = (scrollTop: number, clientHeight: number) => {
      if (scrollTop < clientHeight) setOffsetX(-scrollTop / 10);
    };

    const handleScroll = () => {
      throttle(animatePage(window.pageYOffset, window.innerHeight));
    };

    const event = 'scroll';
    window.addEventListener(event, handleScroll);

    return () => {
      window.removeEventListener(event, handleScroll);
    };
  });

  useResetScrollOnUnmount();

  const renderMilestones = () =>
    Object.values(milestones).map(val => (
      <Milestone
        key={val.title}
        alt={val.alt}
        date={val.date}
        description={val.description}
        photo={val.photo}
        title={val.title}
        url={val.url}
        urlTitle={val.urlTitle}
      />
    ));

  return (
    <div className={styles.Home}>
      <Cover offsetX={offsetX} />
      {renderMilestones()}
    </div>
  );
};

export default Home;
