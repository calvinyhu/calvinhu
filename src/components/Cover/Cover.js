import React from 'react';
import throttle from 'raf-throttle';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import styles from './Cover.module.scss';
import { PAGE } from '../../containers/Home/Home';
import Button from '../UI/Button/Button';
import Fa from '../UI/Icon/Fa/Fa';
import p1080 from '../../assets/images/DSC_9569-1080p50-blurred.webp';
import p1440 from '../../assets/images/DSC_9569-1440p50-blurred.webp';
import p2160 from '../../assets/images/DSC_9569-2160p35-blurred.webp';

const FRICTION = 1 / 30;
const DISPLACE_X = 20;
const DISPLACE_Y = 20;

class Cover extends React.PureComponent {
  static propTypes = {
    offsetX: PropTypes.number.isRequired,
    click: PropTypes.func.isRequired
  };

  state = {
    x: 0,
    y: 0,
    followX: 0,
    followY: 0,
    isAnimating: false
  };

  clickWeb = () => this.props.click(PAGE.WEB);
  clickAbout = () => this.props.click(PAGE.ABOUT);
  clickResume = () => this.props.click(PAGE.RESUME);

  handleMouseMove = event => {
    if (event.movementX !== 0 || event.movementY !== 0)
      throttle(this.setAnimationValues(event.clientX, event.clientY));
  };

  // Effect provided by Fabio Ottaviani (https://codepen.io/supah/pen/RrzREx) that I modified to not run infinitely
  setAnimationValues = (mouseX, mouseY) => {
    if (!mouseX && !mouseY) return;

    let x = null;
    let y = null;
    let followX = null;
    let followY = null;

    x = Math.max(-100, Math.min(100, window.innerWidth / 2 - mouseX));
    y = Math.max(-100, Math.min(100, window.innerHeight / 2 - mouseY));
    followX = (DISPLACE_X * x) / 100;
    followY = (DISPLACE_Y * y) / 100;

    this.setState({ followX: followX, followY: followY });

    // Start the animation loop
    if (!this.state.isAnimating) {
      this.setState({ isAnimating: true });
      this.animateBackground();
    }
  };

  // Effect provided by Fabio Ottaviani (https://codepen.io/supah/pen/RrzREx) that I modified to not run infinitely
  animateBackground = () => {
    if (window.location.pathname !== '/') return;

    this.setState(prevState => {
      return {
        x: prevState.x + (this.state.followX - prevState.x) * FRICTION,
        y: prevState.y + (this.state.followY - prevState.y) * FRICTION
      };
    });

    const stateX = Math.abs(this.state.x);
    const stateY = Math.abs(this.state.y);
    const followX = Math.abs(this.state.followX);
    const followY = Math.abs(this.state.followY);

    // Only continue the animation if within a 1px threshold OR initial animation
    if (stateX === 0 || stateX < followX - 1 || stateY < followY - 1)
      window.requestAnimationFrame(this.animateBackground);
    else this.setState({ isAnimating: false });
  };

  render() {
    const lightenFilter = <div className={styles.LightenFilter} />;
    const background = (
      <div
        style={{
          transform: `translate3d(${this.state.x + this.props.offsetX}px, ${
            this.state.y
          }px, 0px) scale3d(1.4,1.4,1.4)`
        }}
        className={styles.Background}
      >
        <picture>
          <source media="(min-width: 1440px)" srcSet={p1440} />
          <source media="(min-width: 2160px)" srcSet={p2160} />
          <img src={p1080} alt="me" />
        </picture>
        {lightenFilter}
      </div>
    );

    const blurb = (
      <div className={styles.Blurb}>
        <h1>
          Hi, I'm <span>Calvin Hu</span>.
        </h1>
        <p>I'm a web developer.</p>
      </div>
    );

    const nav = (
      <div className={styles.Nav}>
        <div className={styles.PageLinks}>
          <div className={styles.PageLink}>
            <Button link click={this.clickAbout}>
              ABOUT
            </Button>
          </div>
          <div className={styles.PageLink}>
            <Button link click={this.clickWeb}>
              PORTFOLIO
            </Button>
          </div>
          <div className={styles.PageLink}>
            <Button link click={this.clickResume}>
              RESUME
            </Button>
          </div>
        </div>
      </div>
    );

    const socialMedia = (
      <div className={styles.SocialMedia}>
        <a
          href="https://www.linkedin.com/in/calvinyhu/"
          target="_blank"
          rel="noopener noreferrer"
        >
          <Fa twoX>fab fa-linkedin</Fa>
        </a>
        <a
          href="https://www.github.com/calvinyhu/"
          target="_blank"
          rel="noopener noreferrer"
        >
          <Fa twoX>fab fa-github</Fa>
        </a>
        <Link to="/photo">
          <Fa twoX>fas fa-camera-retro</Fa>
        </Link>
      </div>
    );

    const coverText = (
      <div className={styles.CoverText}>
        {blurb}
        {nav}
      </div>
    );

    const coverContent = (
      <div className={styles.CoverContent}>
        {coverText}
        {socialMedia}
      </div>
    );

    return (
      <div className={styles.Cover} onMouseMove={this.handleMouseMove}>
        {background}
        {coverContent}
      </div>
    );
  }
}

export default Cover;
