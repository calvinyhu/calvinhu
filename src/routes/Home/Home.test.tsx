import React from 'react';
import { shallow } from 'enzyme';

import Home from './Home';
import Cover from 'components/Cover/Cover';
import Milestone from 'components/Milestone/Milestone';

describe('<Home />', () => {
  it('should render <Cover />', () => {
    const wrapper = shallow(<Home />);
    expect(wrapper.find(Cover).length).toEqual(1);
  });

  it('should render <Milestone />', () => {
    const wrapper = shallow(<Home />);
    expect(wrapper.find(Milestone).length).toEqual(3);
  });
});
