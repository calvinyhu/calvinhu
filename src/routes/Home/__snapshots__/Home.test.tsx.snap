// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> snapshots should match default snapshot: default 1`] = `ShallowWrapper {}`;
